File: Blink/Blink/NetworkProtocol.swift
//
//  NetworkProtocol.swift
//  Blink
//
//  Created by Zaki Sheriff on 2025-11-26.
//

import Foundation

// MARK: - Network Protocol Constants
enum NetworkConstants {
    static let discoveryPort: UInt16 = 8888
    static let transferPort: UInt16 = 8889
    static let broadcastInterval: TimeInterval = 2.0
    static let deviceTimeout: TimeInterval = 10.0
    static let bufferSize: Int = 65536 // 64KB chunks
    static let protocolVersion: String = "1.0"
}

// MARK: - Message Types
enum MessageType: String, Codable {
    case discovery = "DISCOVERY"
    case discoveryResponse = "DISCOVERY_RESPONSE"
    case pairRequest = "PAIR_REQUEST"
    case pairResponse = "PAIR_RESPONSE"
    case transferRequest = "TRANSFER_REQUEST"
    case transferResponse = "TRANSFER_RESPONSE"
    case transferData = "TRANSFER_DATA"
    case transferComplete = "TRANSFER_COMPLETE"
    case transferError = "TRANSFER_ERROR"
    case progress = "PROGRESS"
    case ping = "PING"
    case pong = "PONG"
}

// MARK: - Protocol Message
struct ProtocolMessage: Codable {
    let type: MessageType
    let timestamp: TimeInterval
    let payload: Data?
    
    init(type: MessageType, payload: Data? = nil) {
        self.type = type
        self.timestamp = Date().timeIntervalSince1970
        self.payload = payload
    }
    
    func toData() -> Data? {
        return try? JSONEncoder().encode(self)
    }
    
    static func fromData(_ data: Data) -> ProtocolMessage? {
        return try? JSONDecoder().decode(ProtocolMessage.self, from: data)
    }
}

// MARK: - Discovery Message
struct DiscoveryMessage: Codable {
    let deviceId: String
    let deviceName: String
    let deviceType: String
    let port: Int
    let protocolVersion: String
    let code: String?
    let type: String // "presence", "query", "response"
    
    init(deviceId: String, deviceName: String, deviceType: String = "macOS", port: Int, code: String? = nil, type: String = "presence") {
        self.deviceId = deviceId
        self.deviceName = deviceName
        self.deviceType = deviceType
        self.port = port
        self.protocolVersion = NetworkConstants.protocolVersion
        self.code = code
        self.type = type
    }
}

// MARK: - Pair Request
struct PairRequest: Codable {
    let deviceId: String
    let deviceName: String
    let timestamp: TimeInterval
}

// MARK: - Pair Response
struct PairResponse: Codable {
    let accepted: Bool
    let deviceId: String
    let deviceName: String
    let sessionToken: String?
}

// MARK: - Transfer Request
struct TransferRequest: Codable {
    let transferId: String
    let fileName: String
    let fileSize: Int64
    let isDirectory: Bool
    let checksum: String?
}

// MARK: - Transfer Response
struct TransferResponse: Codable {
    let transferId: String
    let accepted: Bool
    let reason: String?
}

// MARK: - Transfer Data Packet
struct TransferDataPacket {
    let transferId: String
    let sequenceNumber: Int64
    let data: Data
    let isLastPacket: Bool
    
    func toData() -> Data {
        var result = Data()
        
        // Transfer ID (36 bytes for UUID string)
        if let idData = transferId.data(using: .utf8) {
            var idBytes = [UInt8](repeating: 0, count: 36)
            let copyCount = min(idData.count, 36)
            idData.copyBytes(to: &idBytes, count: copyCount)
            result.append(contentsOf: idBytes)
        }
        
        // Sequence number (8 bytes)
        var seq = sequenceNumber.bigEndian
        result.append(Data(bytes: &seq, count: 8))
        
        // Is last packet flag (1 byte)
        result.append(isLastPacket ? 1 : 0)
        
        // Data length (4 bytes)
        var length = UInt32(data.count).bigEndian
        result.append(Data(bytes: &length, count: 4))
        
        // Actual data
        result.append(data)
        
        return result
    }
    
    static func fromData(_ data: Data) -> TransferDataPacket? {
        guard data.count >= 49 else { return nil } // Minimum size
        
        var offset = 0
        
        // Transfer ID
        let idData = data.subdata(in: offset..<(offset + 36))
        guard let transferId = String(data: idData, encoding: .utf8)?.trimmingCharacters(in: .controlCharacters) else {
            return nil
        }
        offset += 36
        
        // Sequence number
        let seqData = data.subdata(in: offset..<(offset + 8))
        let sequenceNumber = seqData.withUnsafeBytes { $0.load(as: Int64.self) }.bigEndian
        offset += 8
        
        // Is last packet
        let isLastPacket = data[offset] == 1
        offset += 1
        
        // Data length
        let lengthData = data.subdata(in: offset..<(offset + 4))
        let length = lengthData.withUnsafeBytes { $0.load(as: UInt32.self) }.bigEndian
        offset += 4
        
        // Actual data
        guard data.count >= offset + Int(length) else { return nil }
        let packetData = data.subdata(in: offset..<(offset + Int(length)))
        
        return TransferDataPacket(
            transferId: transferId,
            sequenceNumber: sequenceNumber,
            data: packetData,
            isLastPacket: isLastPacket
        )
    }
}

// MARK: - Progress Update
struct ProgressUpdate: Codable {
    let transferId: String
    let bytesTransferred: Int64
    let totalBytes: Int64
    let speed: Double
}

// MARK: - Transfer Complete
struct TransferComplete: Codable {
    let transferId: String
    let success: Bool
    let checksum: String?
    let error: String?
}

File: android/app/src/main/java/com/example/blink/network/NetworkProtocol.kt
package com.example.blink.network

// MARK: - Network Protocol Constants
object NetworkConstants {
    const val DISCOVERY_PORT = 8888
    const val TRANSFER_PORT = 8889
    const val BROADCAST_INTERVAL = 2000L // milliseconds
    const val DEVICE_TIMEOUT = 10000L // milliseconds
    const val BUFFER_SIZE = 65536 // 64KB chunks
    const val PROTOCOL_VERSION = "1.0"
}

// MARK: - Message Types
enum class MessageType {
    DISCOVERY,
    DISCOVERY_RESPONSE,
    PAIR_REQUEST,
    PAIR_RESPONSE,
    TRANSFER_REQUEST,
    TRANSFER_RESPONSE,
    TRANSFER_DATA,
    TRANSFER_COMPLETE,
    TRANSFER_ERROR,
    PROGRESS,
    PING,
    PONG
}

// MARK: - Protocol Message
data class ProtocolMessage(
        val type: MessageType,
        val timestamp: Long = System.currentTimeMillis(),
        val payload: String? = null
)

// MARK: - Discovery Message
data class DiscoveryMessage(
        val deviceId: String,
        val deviceName: String,
        val deviceType: String = "android",
        val port: Int,
        val protocolVersion: String = NetworkConstants.PROTOCOL_VERSION,
        val code: String? = null,
        val type: String = "presence" // "presence", "query", "response"
)

// MARK: - Pair Request
data class PairRequest(
        val deviceId: String,
        val deviceName: String,
        val timestamp: Long = System.currentTimeMillis()
)

// MARK: - Pair Response
data class PairResponse(
        val accepted: Boolean,
        val deviceId: String,
        val deviceName: String,
        val sessionToken: String? = null
)

// MARK: - Transfer Request
data class TransferRequest(
        val transferId: String,
        val fileName: String,
        val fileSize: Long,
        val isDirectory: Boolean,
        val checksum: String? = null
)

// MARK: - Transfer Response
data class TransferResponse(
        val transferId: String,
        val accepted: Boolean,
        val reason: String? = null
)

// MARK: - Transfer Data Packet
data class TransferDataPacket(
        val transferId: String,
        val sequenceNumber: Long,
        val data: ByteArray,
        val isLastPacket: Boolean
) {
    fun toBytes(): ByteArray {
        val result = mutableListOf<Byte>()

        // Transfer ID (36 bytes for UUID string)
        val idBytes = transferId.toByteArray().take(36)
        result.addAll(idBytes)
        result.addAll(List(36 - idBytes.size) { 0.toByte() })

        // Sequence number (8 bytes)
        val seqBytes = ByteArray(8)
        for (i in 0..7) {
            seqBytes[7 - i] = (sequenceNumber shr (i * 8)).toByte()
        }
        result.addAll(seqBytes.toList())

        // Is last packet flag (1 byte)
        result.add(if (isLastPacket) 1 else 0)

        // Data length (4 bytes)
        val length = data.size
        val lengthBytes = ByteArray(4)
        for (i in 0..3) {
            lengthBytes[3 - i] = (length shr (i * 8)).toByte()
        }
        result.addAll(lengthBytes.toList())

        // Actual data
        result.addAll(data.toList())

        return result.toByteArray()
    }

    companion object {
        fun fromBytes(bytes: ByteArray): TransferDataPacket? {
            if (bytes.size < 49) return null // Minimum size

            var offset = 0

            // Transfer ID
            val transferId = String(bytes.sliceArray(offset until offset + 36)).trim('\u0000')
            offset += 36

            // Sequence number
            var sequenceNumber = 0L
            for (i in 0..7) {
                sequenceNumber = (sequenceNumber shl 8) or (bytes[offset + i].toLong() and 0xFF)
            }
            offset += 8

            // Is last packet
            val isLastPacket = bytes[offset] == 1.toByte()
            offset += 1

            // Data length
            var length = 0
            for (i in 0..3) {
                length = (length shl 8) or (bytes[offset + i].toInt() and 0xFF)
            }
            offset += 4

            // Actual data
            if (bytes.size < offset + length) return null
            val data = bytes.sliceArray(offset until offset + length)

            return TransferDataPacket(transferId, sequenceNumber, data, isLastPacket)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TransferDataPacket

        if (transferId != other.transferId) return false
        if (sequenceNumber != other.sequenceNumber) return false
        if (!data.contentEquals(other.data)) return false
        if (isLastPacket != other.isLastPacket) return false

        return true
    }

    override fun hashCode(): Int {
        var result = transferId.hashCode()
        result = 31 * result + sequenceNumber.hashCode()
        result = 31 * result + data.contentHashCode()
        result = 31 * result + isLastPacket.hashCode()
        return result
    }
}

// MARK: - Progress Update
data class ProgressUpdate(
        val transferId: String,
        val bytesTransferred: Long,
        val totalBytes: Long,
        val speed: Double
)

// MARK: - Transfer Complete
data class TransferComplete(
        val transferId: String,
        val success: Boolean,
        val checksum: String? = null,
        val error: String? = null
)
